draft-ietf-mls-protocol.md

```diff
@@ -1388,6 +1400,8 @@ commit_secret -> HKDF-Extract = epoch_secret
                init_secret_[n]
 ~~~~~
 
+[[OPEN ISSUE: `unique_value` needs to be chosen to be derived from information that can be expected to be unique. Choose such information]]
```

Comment:
Is there a reason that `unique_value` needs to be folded in  both at the `epoch_secret` level and for the secrets derived from it?  Seems like it would be simpler to fold it into the `epoch_secret`.

Reply:
Also, "unique" in what sense?

Reply:
Excellent questions! Thanks :-)

**Double use of unique value:**
Derive-Secrets derives a keys from each input keys and xors them and the result is assigned to epoch_secret. It is important that the inputs to the xor have not been used before and the unique value ensures that. tl;dr: 1st use of unique_value is needed for pseudorandomness.

The second use of the unique value ensures that different sessions derive different key values. For honest sessions, this is the case anyway (since pseudorandom key values are unlikely to collide), but for dishonest sessions, the unique value is needed since the xor does not maintain collision-resistance.

**Definition of unique:**
In a call to a Derive-Secrets function with fixed label, e.g.,

Derive-Secrets([., ., .], "combine", unique_value),

the unique-value should not repeat (across protocol runs), even with different input keys. In Derive-Secret with fixed label, e.g., 

Derive-Secret(., "sender data", (unique_value,.)),

the unique-value should not repeat (across protocol runs), even with different input keys. However, it is allowed that unique_values are re-used when the labels are different.

Reply:
Thanks for the clarification on the double input.

I think the GroupContext value does meet the requirement for uniqueness here.  It contains the following values:

* Group ID
* Epoch
* Tree hash
* Transcript hash

Obviously, GroupID and Epoch provide a first level of protection, and the tree hash diverges if there are any different choices of keys. Even if the participants make the same choices in two protocol runs, HPKE uses random keys which are fed into the transcript hash, so will be different.

Reply:
Okay, this sounds very good, thank you!